File: apps/desktop/src/services/widgets.ts
import { listen } from "@tauri-apps/api/event";
import type { WidgetDisplayMode, WidgetSize } from "@tool-center/plugin-contract";

import { hostBridge } from "../runtime/host";
import type { Release } from "@tool-center/plugin-contract";
import { isTauriHost } from "@tool-center/plugin-runtime";

export interface NormalizedPosition {
  readonly x: number;
  readonly y: number;
}

export interface WidgetDimensions {
  readonly width: number;
  readonly height: number;
}

export interface WidgetInstance {
  readonly instanceId: string;
  readonly pluginId: string;
  readonly widgetId: string;
  readonly size: WidgetSize;
  readonly visible: boolean;
  readonly locked: boolean;
  readonly displayMode: WidgetDisplayMode;
  readonly monitorId: string;
  readonly position: NormalizedPosition;
  readonly lastValidPosition: NormalizedPosition;
  readonly dimensions: WidgetDimensions;
  readonly scaleFactor: number;
}

export interface WidgetMonitor {
  readonly id: string;
  readonly name: string;
  readonly primary: boolean;
  readonly scaleFactor: number;
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
}

export interface CreateWidgetInstance {
  readonly pluginId: string;
  readonly widgetId: string;
  readonly size: WidgetSize;
  readonly visible: boolean;
  readonly dimensions: WidgetDimensions;
  readonly monitorId?: string;
}

export interface UpdateWidgetInstance {
  readonly instanceId: string;
  readonly size?: WidgetSize;
  readonly visible?: boolean;
  readonly locked?: boolean;
  readonly displayMode?: WidgetDisplayMode;
  readonly monitorId?: string;
  readonly position?: NormalizedPosition;
  readonly dimensions?: WidgetDimensions;
}

export interface WidgetRegion {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
}

const browserMonitor: WidgetMonitor = {
  id: "browser-primary",
  name: "浏览器预览",
  primary: true,
  scaleFactor: 1,
  x: 0,
  y: 0,
  width: typeof window === "undefined" ? 1280 : window.innerWidth,
  height: typeof window === "undefined" ? 720 : window.innerHeight,
};
let browserInstances: WidgetInstance[] = [];
const browserListeners = new Set<() => void>();

export const widgetService = {
  async list(): Promise<readonly WidgetInstance[]> {
    if (!isTauriHost()) {
      return browserInstances;
    }
    return hostBridge.invoke<readonly WidgetInstance[]>("widget_instances_list");
  },

  async listForHost(
    monitorId: string,
    displayMode: WidgetDisplayMode,
  ): Promise<readonly WidgetInstance[]> {
    if (!isTauriHost()) {
      return browserInstances.filter(
        (instance) =>
          instance.visible &&
          instance.monitorId === monitorId &&
          instance.displayMode === displayMode,
      );
    }
    return hostBridge.invoke<readonly WidgetInstance[]>("widget_host_instances", {
      query: { monitorId, displayMode },
    });
  },

  async monitors(): Promise<readonly WidgetMonitor[]> {
    if (!isTauriHost()) {
      return [{ ...browserMonitor, width: window.innerWidth, height: window.innerHeight }];
    }
    return hostBridge.invoke<readonly WidgetMonitor[]>("widget_monitors_list");
  },

  async create(input: CreateWidgetInstance): Promise<WidgetInstance> {
    if (!isTauriHost()) {
      const position = { x: 0.04 + browserInstances.length * 0.035, y: 0.04 };
      const instance: WidgetInstance = {
        instanceId: `browser-${Date.now()}-${browserInstances.length}`,
        pluginId: input.pluginId,
        widgetId: input.widgetId,
        size: input.size,
        visible: input.visible,
        locked: false,
        displayMode: "desktop",
        monitorId: input.monitorId ?? browserMonitor.id,
        position,
        lastValidPosition: position,
        dimensions: input.dimensions,
        scaleFactor: 1,
      };
      browserInstances = [...browserInstances, instance];
      emitBrowserChange();
      return instance;
    }
    return hostBridge.invoke<WidgetInstance>("widget_instance_create", { input });
  },

  async update(input: UpdateWidgetInstance): Promise<WidgetInstance> {
    if (!isTauriHost()) {
      const index = browserInstances.findIndex(
        (instance) => instance.instanceId === input.instanceId,
      );
      if (index < 0) {
        throw new Error("Widget instance was not found.");
      }
      const previous = browserInstances[index]!;
      const updated: WidgetInstance = {
        ...previous,
        ...input,
        lastValidPosition: input.position ?? previous.lastValidPosition,
      };
      browserInstances = browserInstances.map((instance, candidateIndex) =>
        candidateIndex === index ? updated : instance,
      );
      emitBrowserChange();
      return updated;
    }
    return hostBridge.invoke<WidgetInstance>("widget_instance_update", { input });
  },

  async reorder(instanceIds: readonly string[]): Promise<readonly WidgetInstance[]> {
    if (!isTauriHost()) {
      const byId = new Map(
        browserInstances.map((instance) => [instance.instanceId, instance] as const),
      );
      if (
        instanceIds.length !== browserInstances.length ||
        new Set(instanceIds).size !== browserInstances.length ||
        instanceIds.some((instanceId) => !byId.has(instanceId))
      ) {
        throw new Error("Widget instance order does not match the current instances.");
      }
      browserInstances = instanceIds.map((instanceId) => byId.get(instanceId)!);
      emitBrowserChange();
      return browserInstances;
    }
    return hostBridge.invoke<readonly WidgetInstance[]>("widget_instances_reorder", {
      instanceIds: [...instanceIds],
    });
  },

  async remove(instanceId: string): Promise<void> {
    if (!isTauriHost()) {
      browserInstances = browserInstances.filter(
        (instance) => instance.instanceId !== instanceId,
      );
      emitBrowserChange();
      return;
    }
    await hostBridge.invoke<void>("widget_instance_remove", { instanceId });
  },

  async resetPosition(instanceId: string): Promise<WidgetInstance> {
    if (!isTauriHost()) {
      return this.update({ instanceId, monitorId: browserMonitor.id, position: { x: 0.04, y: 0.04 } });
    }
    return hostBridge.invoke<WidgetInstance>("widget_instance_reset_position", { instanceId });
  },

  async syncHosts(enabledPluginIds: readonly string[]): Promise<void> {
    if (!isTauriHost()) {
      return;
    }
    await hostBridge.invoke<void>("widget_hosts_sync", {
      enabledPluginIds: [...enabledPluginIds],
    });
  },

  async setInteractiveRegions(regions: readonly WidgetRegion[] | null): Promise<void> {
    if (!isTauriHost()) {
      return;
    }
    await hostBridge.invoke<void>("widget_host_set_regions", {
      regions: regions === null ? null : [...regions],
    });
  },

  async subscribeChanges(listener: () => void): Promise<Release> {
    if (!isTauriHost()) {
      browserListeners.add(listener);
      return () => {
        browserListeners.delete(listener);
      };
    }
    return listen("toolcenter://widgets-changed", listener);
  },
};

function emitBrowserChange(): void {
  for (const listener of browserListeners) {
    listener();
  }
}

File: apps/desktop/src-tauri/src/commands/widgets.rs
use std::collections::{HashMap, HashSet};

use serde::Deserialize;
use tauri::{
    AppHandle, Emitter, Manager, PhysicalPosition, PhysicalSize, State, WebviewUrl, WebviewWindow,
    WebviewWindowBuilder,
};

use crate::error::AppError;
use crate::services::widgets::{
    self, CreateWidgetInstance, UpdateWidgetInstance, WidgetDisplayMode, WidgetHostState,
    WidgetInstance, WidgetMonitor,
};
use crate::state::CoreState;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HostQuery {
    monitor_id: String,
    display_mode: WidgetDisplayMode,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WidgetRegion {
    x: f64,
    y: f64,
    width: f64,
    height: f64,
    #[serde(default)]
    radius: f64,
}

#[tauri::command]
pub fn widget_instances_list(state: State<'_, CoreState>) -> Result<Vec<WidgetInstance>, AppError> {
    widgets::list(&state)
}

#[tauri::command]
pub fn widget_host_instances(
    state: State<'_, CoreState>,
    host_state: State<'_, WidgetHostState>,
    query: HostQuery,
) -> Result<Vec<WidgetInstance>, AppError> {
    widgets::list_for_host(&state, &host_state, &query.monitor_id, &query.display_mode)
}

#[tauri::command]
pub fn widget_monitors_list(app: AppHandle) -> Result<Vec<WidgetMonitor>, AppError> {
    widgets::monitors(&app)
}

#[tauri::command]
pub async fn widget_instance_create(
    app: AppHandle,
    state: State<'_, CoreState>,
    host_state: State<'_, WidgetHostState>,
    input: CreateWidgetInstance,
) -> Result<WidgetInstance, AppError> {
    let instance = widgets::create(&app, &state, input)?;
    sync_hosts(&app, &state, &host_state)?;
    Ok(instance)
}

#[tauri::command]
pub async fn widget_instance_update(
    app: AppHandle,
    state: State<'_, CoreState>,
    host_state: State<'_, WidgetHostState>,
    input: UpdateWidgetInstance,
) -> Result<WidgetInstance, AppError> {
    let instance = widgets::update(&app, &state, input)?;
    sync_hosts(&app, &state, &host_state)?;
    Ok(instance)
}

#[tauri::command]
pub async fn widget_instances_reorder(
    app: AppHandle,
    state: State<'_, CoreState>,
    host_state: State<'_, WidgetHostState>,
    instance_ids: Vec<String>,
) -> Result<Vec<WidgetInstance>, AppError> {
    let instances = widgets::reorder(&state, &instance_ids)?;
    sync_hosts(&app, &state, &host_state)?;
    Ok(instances)
}

#[tauri::command]
pub async fn widget_instance_remove(
    app: AppHandle,
    state: State<'_, CoreState>,
    host_state: State<'_, WidgetHostState>,
    instance_id: String,
) -> Result<(), AppError> {
    widgets::remove(&state, &instance_id)?;
    sync_hosts(&app, &state, &host_state)
}

#[tauri::command]
pub async fn widget_instance_reset_position(
    app: AppHandle,
    state: State<'_, CoreState>,
    host_state: State<'_, WidgetHostState>,
    instance_id: String,
) -> Result<WidgetInstance, AppError> {
    let instance = widgets::reset_position(&app, &state, &instance_id)?;
    sync_hosts(&app, &state, &host_state)?;
    Ok(instance)
}

#[tauri::command]
pub async fn widget_hosts_sync(
    app: AppHandle,
    state: State<'_, CoreState>,
    host_state: State<'_, WidgetHostState>,
    enabled_plugin_ids: Vec<String>,
) -> Result<(), AppError> {
    host_state.set_enabled(&enabled_plugin_ids)?;
    sync_hosts(&app, &state, &host_state)
}

#[tauri::command]
pub fn widget_host_set_regions(
    window: WebviewWindow,
    regions: Option<Vec<WidgetRegion>>,
) -> Result<(), AppError> {
    if !window.label().starts_with("widget-host-") {
        return Err(AppError::invalid_input(
            "Only a widget host can update its interactive region.",
        ));
    }
    set_window_regions(&window, regions.as_deref())
}

fn sync_hosts(
    app: &AppHandle,
    state: &CoreState,
    host_state: &WidgetHostState,
) -> Result<(), AppError> {
    let instances = widgets::reconcile(app, state)?;
    let monitors = widgets::monitors(app)?;
    let monitor_map = monitors
        .iter()
        .map(|monitor| (monitor.id.as_str(), monitor))
        .collect::<HashMap<_, _>>();
    let mut desired = HashSet::new();

    for instance in instances.iter().filter(|instance| {
        instance.visible && host_state.is_enabled(&instance.plugin_id).unwrap_or(false)
    }) {
        let Some(monitor) = monitor_map.get(instance.monitor_id.as_str()) else {
            continue;
        };
        let label = widgets::host_label(&instance.monitor_id, &instance.display_mode);
        desired.insert(label.clone());
        ensure_host_window(app, &label, monitor, &instance.display_mode)?;
    }

    for (label, window) in app.webview_windows() {
        if !label.starts_with("widget-host-") {
            continue;
        }
        if desired.contains(&label) {
            window
                .emit("toolcenter://widgets-changed", ())
                .map_err(|error| AppError::new("widgets.emit-failed", error.to_string()))?;
        } else {
            window
                .close()
                .map_err(|error| AppError::new("widgets.window-failed", error.to_string()))?;
        }
    }
    Ok(())
}

fn ensure_host_window(
    app: &AppHandle,
    label: &str,
    monitor: &WidgetMonitor,
    display_mode: &WidgetDisplayMode,
) -> Result<(), AppError> {
    if let Some(window) = app.get_webview_window(label) {
        configure_host_window(&window, monitor, display_mode)?;
        return Ok(());
    }

    let url = format!(
        "index.html?surface=widget-host&monitorId={}&displayMode={}",
        monitor.id,
        match display_mode {
            WidgetDisplayMode::Desktop => "desktop",
            WidgetDisplayMode::AlwaysOnTop => "always-on-top",
        }
    );
    let window = WebviewWindowBuilder::new(app, label, WebviewUrl::App(url.into()))
        .title("ToolCenter Widget Host")
        .decorations(false)
        .transparent(true)
        .shadow(false)
        .skip_taskbar(true)
        .resizable(false)
        .always_on_top(matches!(display_mode, WidgetDisplayMode::AlwaysOnTop))
        .focused(false)
        .visible(false)
        .inner_size(200.0, 120.0)
        .build()
        .map_err(|error| AppError::new("widgets.window-failed", error.to_string()))?;
    configure_host_window(&window, monitor, display_mode)?;
    set_window_regions(&window, Some(&[]))?;
    window
        .show()
        .map_err(|error| AppError::new("widgets.window-failed", error.to_string()))
}

fn configure_host_window(
    window: &WebviewWindow,
    monitor: &WidgetMonitor,
    display_mode: &WidgetDisplayMode,
) -> Result<(), AppError> {
    window
        .set_position(PhysicalPosition::new(monitor.x, monitor.y))
        .map_err(|error| AppError::new("widgets.window-failed", error.to_string()))?;
    window
        .set_size(PhysicalSize::new(monitor.width, monitor.height))
        .map_err(|error| AppError::new("widgets.window-failed", error.to_string()))?;
    window
        .set_always_on_top(matches!(display_mode, WidgetDisplayMode::AlwaysOnTop))
        .map_err(|error| AppError::new("widgets.window-failed", error.to_string()))
}

#[cfg(windows)]
fn set_window_regions(
    window: &WebviewWindow,
    regions: Option<&[WidgetRegion]>,
) -> Result<(), AppError> {
    use windows::Win32::Graphics::Gdi::{
        CombineRgn, CreateRectRgn, CreateRoundRectRgn, DeleteObject, HGDIOBJ, RGN_OR, SetWindowRgn,
    };

    let hwnd = window
        .hwnd()
        .map_err(|error| AppError::new("widgets.window-failed", error.to_string()))?;
    let Some(regions) = regions else {
        // SAFETY: The HWND belongs to this Tauri window. A null region restores the full window.
        unsafe { SetWindowRgn(hwnd, None, true) };
        return Ok(());
    };
    let scale = window
        .scale_factor()
        .map_err(|error| AppError::new("widgets.window-failed", error.to_string()))?;
    // SAFETY: GDI region handles are created and released according to SetWindowRgn ownership rules.
    unsafe {
        let combined = CreateRectRgn(0, 0, 0, 0);
        if combined.is_invalid() {
            return Err(AppError::new(
                "widgets.region-failed",
                "Unable to create the widget hit-test region.",
            ));
        }
        for region in regions {
            if !region.x.is_finite()
                || !region.y.is_finite()
                || !region.width.is_finite()
                || !region.height.is_finite()
                || !region.radius.is_finite()
                || region.width <= 0.0
                || region.height <= 0.0
                || region.radius < 0.0
            {
                let _ = DeleteObject(HGDIOBJ(combined.0));
                return Err(AppError::invalid_input("Widget region is invalid."));
            }
            let left = (region.x * scale).round() as i32;
            let top = (region.y * scale).round() as i32;
            let right = ((region.x + region.width) * scale).round() as i32;
            let bottom = ((region.y + region.height) * scale).round() as i32;
            let corner_radius = region
                .radius
                .min(region.width / 2.0)
                .min(region.height / 2.0);
            let corner_diameter = (corner_radius * 2.0 * scale).round() as i32;
            let widget_region = if corner_diameter > 0 {
                CreateRoundRectRgn(left, top, right, bottom, corner_diameter, corner_diameter)
            } else {
                CreateRectRgn(left, top, right, bottom)
            };
            if widget_region.is_invalid() {
                let _ = DeleteObject(HGDIOBJ(combined.0));
                return Err(AppError::new(
                    "widgets.region-failed",
                    "Unable to create a widget hit-test region.",
                ));
            }
            let _ = CombineRgn(Some(combined), Some(combined), Some(widget_region), RGN_OR);
            let _ = DeleteObject(HGDIOBJ(widget_region.0));
        }
        if SetWindowRgn(hwnd, Some(combined), true) == 0 {
            let _ = DeleteObject(HGDIOBJ(combined.0));
            return Err(AppError::new(
                "widgets.region-failed",
                "Unable to apply the widget hit-test region.",
            ));
        }
    }
    Ok(())
}

#[cfg(not(windows))]
fn set_window_regions(
    _window: &WebviewWindow,
    _regions: Option<&[WidgetRegion]>,
) -> Result<(), AppError> {
    Ok(())
}
